### ISSUE TITLE
Option for full-page-load navigation (if the numbers support it)

### ISSUE THREAD

**Issue by boutell** [vuejs/vitepress#2539]
### Is your feature request related to a problem? Please describe.

Vitepress is a fantastic project, and I apologize for nitpicking. However I'm opening this ticket to see if it's certain that full-page-load navigation via JavaScript rather than loading a new page (given everything already in the browser cache) is really a win on the median device.

On a powerful device, I'll assume it's definitely a win.

However, globally speaking, the typical desktop and especially typical mobile device are surprisingly limited:

https://infrequently.org/2022/12/performance-baseline-2023/#devices

So I wanted to start a conversation about whether the no-full-page-load navigation has been performance-tested on limited devices like these, as compared to just letting full-page-load navigation happen with CSS and JS coming from browser cache.

My guess is that this has been benchmarked. But if not, possibly I could attempt it.

I personally really like the idea of no-full-page-load navigation for "ordinary links," but for that very reason I am questioning it to make sure it is helpful to the average person in 2023, and not just developers with nice iPhones and Macs (:

Thanks for your response!

### Describe the solution you'd like

A conversation re: performance of the no-full-page-load technique on typical, somewhat depressing devices, particularly mobile devices.

### Validations

- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Read the [Contributing Guidelines](https://github.com/vuejs/vitepress/blob/main/.github/contributing.md).
- [X] Check that there isn't already an issue that asks for the same feature to avoid creating a duplicate.

**Comment by brc-dd** [comment]
Try benchmarking it.

The main reason on our side was to make it seem like SPA navigation and avoid page reloads. And I don't think that there will be any gain from the network side of things. The way we are doing this should ideally load up slightly less stuff (with lesser request chaining), and won't affect the caching behavior.

Another thing, you can't specify longer cache control age on pages (the *.html files). But you can do so on assets as they have a hash associated with them. This theoretically means the browser will be cache the things for longer if we are doing client-side navigation, as we just fetch those assets, and they will be in cache if no recent deploy has happened.

**Comment by boutell** [comment]
Thanks for responding!

A lot of interesting ins and outs here. As you point out, a JS framework is still being loaded, even if in principle you might not need one at all for some sites. At that point, the JS overhead of swapping pages the JS way is probably lower than rebooting the framework. A better comparison might have to look at use cases where the framework isn't needed at all, e.g. exactly the same nontrivial site built with something Jekyll-ish versus this, or with vitepress lightly patched to do the server side vue HTML generation only and not load the frontend framework (my preference and certainly a whole lot less work than building two sites).

This could take a minute 😆 But it would be a valuable exercise and I've certainly opened my mouth at this point. I'll see what I can pull together.

**Comment by brc-dd** [comment]
That's what we attempted with the [MPA mode](https://vitepress.dev/guide/mpa-mode) (0 client side JS). That does full page navigations as you were suggesting. But it is far from stable. And the default theme won't work properly with it as it needs JS for interactivity.

**Comment by boutell** [comment]
Thanks @brc-dd ! Will at least some navigation work e.g. ordinary links? That would allow for some reasonable testing.

**Comment by brc-dd** [comment]
Yeah those should work. Navbar/sidebar on mobiles will be broken (and collapsing won't work). Search, dark/light toggle, code copy -- things like these won't work as well.
